hooks/proprietary/PASHR.js

```diff
@@ -0,0 +1,144 @@
+------------------------------------------------------------------------------
+Field Number:
+0. hhmmss.sss - UTC time
+1. hhh.hh - Heading in degrees
+2. T - flag to indicate that the Heading is True Heading (i.e. to True North)
+3. rrr.rr - Roll Angle in degrees
+4. ppp.pp - Pitch Angle in degrees
+5. xxx.xx - Heave
+6. a.aaa - Roll Angle Accuracy Estimate (Stdev) in degrees
+7. b.bbb - Pitch Angle Accuracy Estimate (Stdev) in degrees
+8. c.ccc - Heading Angle Accuracy Estimate (Stdev) in degrees
+9. d - Aiding Status
+10. e - IMU Status
+11. hh - Checksum
+*/
+
+function isEmpty(mixed){
+  return ((typeof mixed !== 'string' && typeof mixed !== 'number') || (typeof mixed === 'string' && mixed.trim() === ''))
+}
+
+module.exports = function(input){
+  const { id, sentence, parts, tags } = input;
+
+  const empty = parts.reduce((e, val) => {
+    if( isEmpty(val) ){
+      ++e
+    }
+    return e
+  }, 0);
+
+  if( empty > 3 ){
+    return null
+  }
+
+  const time = parts[0].indexOf('.') === -1 ? parts[0] : parts[0].split('.')[0];
+  const timestamp = utils.timestamp(time, moment.tz('UTC').format('DDMMYY'));
+
+  const GPSQualityFlags = [
+    'no GPS',
+    'All non-RTK fixed integer positions',
+    'RTK fixed integer position'
+  ];
+
+  const INSStatusFlag = [
+    'Pre-Alignment',
+    'Post-Alignment'
+  ];
+
+  const delta = {
+    updates: [
+      {
+        source: tags.source,
+        timestamp: timestamp,
+        values: [
+          {
+            path: 'navigation.attitude',
+            value: {
+              heading: Number(parts[1]),
+              roll: Number(parts[3]),
+              pitch: Number(parts[4])
+            }
+          },
+          {
+            path: 'navigation.attitude.accuracy',
+            value: {
+              heading: Number(parts[8]),
+              roll: Number(parts[6]),
+              pitch: Number(parts[7])
+            }
+          },
+          {
+            path: 'navigation.attitude.heading',
+            value: {
+              heading: Number(parts[1]),
+              true: parts[2] === "T"
+            }
+          },
+          {
+            path: 'navigation.attitude.heading.true',
+            value: parts[2] === "T"
+          },
+          {
+            path: 'navigation.attitude.heave',
+            value: Number(parts[5])
+          },
+          {
+            path: 'navigation.attitude.aid',
+            value: GPSQualityFlags[utils.int(parts[9])]
+          },
+
+          {
+            path: 'navigation.attitude.status',
+            value: INSStatusFlag[utils.int(parts[10])]
+          },
+        ]
+      }
+    ],
+  };
+
+  const toRemove = [];
+
+  delta.updates[0].values.forEach((update, index) => {
+    if( typeof update.value === 'undefined' || update.value === null || (typeof update.value === 'string' && update.value.trim() === '') || (typeof update.value === 'number' && isNaN(update.value)) ){
```

Comment:
I'd prefer a function that is named so that it descibes the check - this is string of comparisons is hard to decipher. What is it checking?

Imho creating first an array, then a second array of things to remove and then removing them on enyb one is pretty awkward. How about `arr = arr.filter(condition)`?